http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java

```diff
@@ -0,0 +1,349 @@
+        log.debug(() -> "Creating ConnectionPool for: URI:" + uri + ", MaxConns: " + maxConnectionsPerEndpoint);
+        return new HttpConnectionPoolManager(bootstrap, socketOptions, tlsContext, uri, windowSize, maxConnectionsPerEndpoint);
+    }
+
+    private HttpConnectionPoolManager getOrCreateConnectionPool(URI uri) {
+        Validate.notNull(uri, "URI must not be null");
+        HttpConnectionPoolManager connPool = connectionPools.get(uri);
+
+        if (connPool == null) {
+            HttpConnectionPoolManager newConnPool = createConnectionPool(uri);
+            HttpConnectionPoolManager alreadyExistingConnPool = connectionPools.putIfAbsent(uri, newConnPool);
+
+            if (alreadyExistingConnPool == null) {
+                connPool = newConnPool;
+            } else {
+                // Multiple threads trying to open connections to the same URI at once, close the newer one
+                newConnPool.close();
+                connPool = alreadyExistingConnPool;
+            }
+        }
+
+        return connPool;
+    }
+
+    private List<HttpHeader> createHttpHeaderList(URI uri, AsyncExecuteRequest asyncRequest) {
+        SdkHttpRequest sdkRequest = asyncRequest.request();
+        List<HttpHeader> crtHeaderList = new ArrayList<>(sdkRequest.headers().size() + 2);
+
+        // Set Host Header if needed
+        if (isNullOrEmpty(sdkRequest.headers().get(HOST_HEADER))) {
+            crtHeaderList.add(new HttpHeader(HOST_HEADER, uri.getHost()));
+        }
+
+        // Add Connection Keep Alive Header to reuse this Http Connection as long as possible
+        if (isNullOrEmpty(sdkRequest.headers().get(CONNECTION))) {
+            crtHeaderList.add(new HttpHeader(CONNECTION, KEEP_ALIVE));
+        }
+
+        // Set Content-Length if needed
+        Optional<Long> contentLength = asyncRequest.requestContentPublisher().contentLength();
+        if (isNullOrEmpty(sdkRequest.headers().get(CONTENT_LENGTH)) && contentLength.isPresent()) {
+            crtHeaderList.add(new HttpHeader(CONTENT_LENGTH, Long.toString(contentLength.get())));
+        }
+
+        // Add the rest of the Headers
+        for (Map.Entry<String, List<String>> headerList: sdkRequest.headers().entrySet()) {
+            for (String val: headerList.getValue()) {
+                HttpHeader h = new HttpHeader(headerList.getKey(), val);
+                crtHeaderList.add(h);
+            }
+        }
+
+        return crtHeaderList;
+    }
+
+    private HttpHeader[] asArray(List<HttpHeader> crtHeaderList) {
+        return crtHeaderList.toArray(new HttpHeader[crtHeaderList.size()]);
+    }
+
+    private HttpRequest toCrtRequest(URI uri, AsyncExecuteRequest asyncRequest) {
+        SdkHttpRequest sdkRequest = asyncRequest.request();
+        Validate.notNull(uri, "URI must not be null");
+        Validate.notNull(sdkRequest, "SdkHttpRequest must not be null");
+
+        String method = sdkRequest.method().name();
```

Comment:
do we need toUpper() here?

Reply:
`.method()` returns a [SdkHttpMethod Enum](https://github.com/aws/aws-sdk-java-v2/blob/master/http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpMethod.java#L28), and all the Enum value names are Uppercase, so calling `.name()` will always return upper case values.